hathor/indexes/manager.py

```diff
@@ -66,23 +110,82 @@ def enable_tokens_index(self) -> None:
         raise NotImplementedError
 
-    def _manually_initialize_tips_indexes(self, tx_storage: 'TransactionStorage') -> None:
-        """ Initialize the tips indexes, populating them from a tx_storage that is otherwise complete.
-
-        XXX: this method requires timestamp indexes to be complete and up-to-date with the rest of the database
-        XXX: this method is not yet being used
+    def _manually_initialize(self, tx_storage: 'TransactionStorage') -> None:
+        """ Initialize the indexes, checking the indexes that need initialization, and the optimal iterator to use.
         """
-        for tx in tx_storage._topological_sort_timestamp_index():
-            tx_meta = tx.get_metadata()
-            if not tx_meta.validation.is_final():
-                continue
+        from hathor.transaction.genesis import BLOCK_GENESIS
 
-            self.all_tips.add_tx(tx)
+        db_last_started_at = tx_storage.get_last_started_at()
+        sorted_all_db_name = self.sorted_all.get_db_name()
+        if sorted_all_db_name is None:
+            can_use_timestamp_index = False
+        else:
+            sorted_all_index_last_started_at = tx_storage.get_index_last_started_at(sorted_all_db_name)
+            can_use_timestamp_index = db_last_started_at == sorted_all_index_last_started_at
+
+        indexes_to_init: List[Tuple[_IndexFilter, BaseIndex]] = []
+        for index_filter, index in self._iter_all_indexes_with_filter():
+            index_db_name = index.get_db_name()
+            if index_db_name is None:
+                indexes_to_init.append((index_filter, index))
+                continue
+            index_last_started_at = tx_storage.get_index_last_started_at(index_db_name)
+            if db_last_started_at != index_last_started_at:
+                indexes_to_init.append((index_filter, index))
 
+        if indexes_to_init:
+            self.log.debug('there are indexes that need initialization',
+                           indexes_to_init=[i for _, i in indexes_to_init])
+        else:
+            self.log.debug('there are no indexes that need initialization')
+
+        iter_tx: Iterator[BaseTransaction]
+        if can_use_timestamp_index:
+            iter_tx = tx_storage._topological_sort_timestamp_index()
+        # XXX: this mostly exist because some tests will break when not using a caching-storage because of metadata
+        #      objects being duplicated which make the topological metadata iterator not work
+        elif tx_storage.is_caching_storage():
+            iter_tx = tx_storage._topological_sort_metadata()
+        else:
+            iter_tx = tx_storage._topological_sort_dfs()
+
+        # make sure that all the indexes that we're rebuilding are cleared
+        for _, index in indexes_to_init:
+            index_db_name = index.get_db_name()
+            if index_db_name:
+                tx_storage.set_index_last_started_at(index_db_name, 0)
```

Comment:
Why `0`? What does it mean?

Reply:
`0` is the smallest timestamp, it's also not a valid timestamp, it's used to indicate that the index is invalid, it also will never match the database timestamp because we reset that to `1` instead. I'll move these two values (`0` and `1`) to a module constant to give it more meaning and add an explanation.

Reply:
Done.

Reply:
Thanks!